torch/csrc/jit/script/compiler.cpp

```diff
@@ -1486,13 +1494,96 @@ void defineMethodsInModule(Module & m, const std::vector<TypedDef>& definitions,
   }
 }
 
+namespace type_annotations {
+
+std::unordered_map<std::string, TypePtr> ident_to_type_lut() {
+  static std::unordered_map<std::string, TypePtr> map = {
+    {"Tensor", DynamicType::get()},
+    {"Int", IntType::get()},
+    {"Float", FloatType::get()},
+  };
+  return map;
+}
+
+std::unordered_set<std::string> subscriptable_types() {
+  static std::unordered_set<std::string> map = {
+    "Tuple"
+  };
+  return map;
+}
+
+TypePtr parseTypeFromExpr(Expr expr) {
+  if (expr.kind() == TK_VAR) {
+    auto ident = Var(expr).name();
+    if (ident_to_type_lut().count(ident.name())) {
+      return ident_to_type_lut()[ident.name()];
+    }
+    throw ErrorReport(expr.range()) << "Unknown type name " << ident.name();
+  } else if (expr.kind() == TK_SUBSCRIPT) {
+    auto subscript = Subscript(expr);
+    if (subscript.value().kind() != TK_VAR) {
+      throw ErrorReport(subscript.value().range()) << "Subscripted type must be a type identifier";
+    }
+    auto value_name = Var(subscript.value()).name().name();
+    if (!subscriptable_types().count(value_name)) {
```

Comment:
nit: if you are going to just test each valid value_name, there is no need to also have a separate table that has to stay in sync with the if statement chain here.